Blocks not fully working on M1 hardware

**Describe the bug**
The Rubicon ObjC Test suite has failures when executed on M1 hardware. The failures are all in the tests of Block functionality; some, but not all tests are currently failing.

The test failures all appear to be related to the return values from block methods.

**To Reproduce**
Steps to reproduce the behavior:
1. Run the test suite on M1 hardware2. 
2. Test failures occur:
```
==============================================================================
FAIL: tests/test_blocks.py::BlockTests::test_block_bound_method
------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/rkm/beeware/rubicon/objc/tests/test_blocks.py", line 158, in test_block_bound_method
    self.assertEqual(returned_two_args_block(12, 34), 46)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/unittest/case.py", line 912, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/unittest/case.py", line 905, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: None != 46

==============================================================================
FAIL: tests/test_blocks.py::BlockTests::test_block_delegate_auto
------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/rkm/beeware/rubicon/objc/tests/test_blocks.py", line 53, in test_block_delegate_auto
    self.assertEqual(result, 9)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/unittest/case.py", line 912, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/unittest/case.py", line 905, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: 0 != 9

==============================================================================
FAIL: tests/test_blocks.py::BlockTests::test_block_delegate_auto_struct
------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/rkm/beeware/rubicon/objc/tests/test_blocks.py", line 70, in test_block_delegate_auto_struct
    self.assertEqual(result, 85)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/unittest/case.py", line 912, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/unittest/case.py", line 905, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: 0 != 85

==============================================================================
FAIL: tests/test_blocks.py::BlockTests::test_block_delegate_method_manual_ctypes
------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/rkm/beeware/rubicon/objc/tests/test_blocks.py", line 31, in test_block_delegate_method_manual_ctypes
    self.assertEqual(result, 5)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/unittest/case.py", line 912, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/unittest/case.py", line 905, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: 0 != 5

==============================================================================
FAIL: tests/test_blocks.py::BlockTests::test_block_delegate_method_manual_pytypes
------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/rkm/beeware/rubicon/objc/tests/test_blocks.py", line 42, in test_block_delegate_method_manual_pytypes
    self.assertEqual(result, 5)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/unittest/case.py", line 912, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/unittest/case.py", line 905, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: 0 != 5

==============================================================================
FAIL: tests/test_blocks.py::BlockTests::test_block_property_ctypes
------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/rkm/beeware/rubicon/objc/tests/test_blocks.py", line 14, in test_block_property_ctypes
    self.assertEqual(result, 3)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/unittest/case.py", line 912, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/unittest/case.py", line 905, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: None != 3

==============================================================================
FAIL: tests/test_blocks.py::BlockTests::test_block_property_pytypes
------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/rkm/beeware/rubicon/objc/tests/test_blocks.py", line 20, in test_block_property_pytypes
    self.assertEqual(result, 3)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/unittest/case.py", line 912, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/unittest/case.py", line 905, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: None != 3

==============================================================================
FAIL: tests/test_blocks.py::BlockTests::test_block_round_trip
------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/rkm/beeware/rubicon/objc/tests/test_blocks.py", line 118, in test_block_round_trip
    self.assertEqual(returned_block(8, 9), 17)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/unittest/case.py", line 912, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/unittest/case.py", line 905, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: None != 17

------------------------------------------------------------------------------
Ran 266 tests in 2.15s

FAILED (failures=8)
```

**Expected behavior**
Test suite should pass on M1 hardware

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Environment:**
 - M1 Mac hardware
 - Operating System: macOS 12.3.1
 - Python version: 3.8, 3.9, 3.10, 3.11

**Additional context**

It is likely that these failures are also occurring on iOS devices. CI testing this is going to be difficult...

--- a/changes/225.bugfix.rst
+++ b/changes/225.bugfix.rst
@@ -1,0 +1,1 @@
+References to blocks obtained from an Objective C API can now be invoked on M1 hardware.

--- a/changes/225.feature.rst
+++ b/changes/225.feature.rst
@@ -1,0 +1,1 @@
+``objc_id`` and ``objc_block`` are now exposed as part of the ``rubicon.objc`` namespace, rather than requiring an import from ``rubicon.objc.runtime``.

--- a/docs/index.rst
+++ b/docs/index.rst
@@ -40,7 +40,7 @@ Community
 
 Rubicon is part of the `BeeWare suite <https://beeware.org>`__. You can talk to the community through:
 
-* `@pybeeware on Twitter <https://twitter.com/pybeeware>`__
+* `@beeware@fosstodon.org on Mastodon <https://fosstodon.org/@beeware>`__
 
 * `Discord <https://beeware.org/bee/chat/>`__
 

--- a/src/rubicon/objc/__init__.py
+++ b/src/rubicon/objc/__init__.py
@@ -54,7 +54,7 @@ from .api import (
     objc_rawmethod,
     py_from_ns,
 )
-from .runtime import SEL, send_message, send_super
+from .runtime import SEL, objc_block, objc_id, send_message, send_super
 from .types import (
     CFIndex,
     CFRange,
@@ -133,8 +133,10 @@ __all__ = [
     ObjCProtocol,
     at,
     ns_from_py,
+    objc_block,
     objc_classmethod,
     objc_const,
+    objc_id,
     objc_ivar,
     objc_method,
     objc_property,

--- a/src/rubicon/objc/api.py
+++ b/src/rubicon/objc/api.py
@@ -2169,11 +2169,11 @@ class ObjCBlock:
             # first required argument - the block itself.
             block_arg = [objc_id]
 
-        # Setting the restype and argtypes on the invoke function that is in the
-        # ObjCBlockStruct won't stick, because it's not a distinct Python object that
-        # ctypes can use to attach a type hint. Store the ctypes annotations, and apply
-        # them just before invocation. We're going to have to do some light type
-        # conversion in some cases anyway, so this works out well.
+        # If you set restype and argtypes on the invoke function that is in the
+        # ObjCBlockStruct, subsequent gets won't reflect those changes, because it's not
+        # a distinct Python object that ctypes can use to attach a type hint. Store the
+        # ctypes annotations, and apply them just before invocation. We're going to have
+        # to do some light type conversion in some cases anyway, so this works out well.
         self.invoke_restype = ctype_for_type(restype)
         self.invoke_argtypes = block_arg + [
             ctype_for_type(arg_type) for arg_type in argtypes
@@ -2201,13 +2201,21 @@ class ObjCBlock:
         # of type names, so even if the field types of a structure provided as an
         # argument match, ctypes will raise a TypeError.
         #
-        # To avoid this, look for the `__anonymous__` property on the structure
-        # definition. If it exists, check that the provided argument matches the "shape"
-        # of the anonymous structure. If it matches, modify the invoke signature to
-        # match the type of the argument that was actually provided. The first argument
-        # to invoke is the block being invoked, so we can ignore that type hint.
+        # To avoid this, if an argument is a structure, and the argtype for that
+        # argument is a structure, look for the `__anonymous__` property on the argtype
+        # structure definition - this property is added automatically to structures when
+        # a structure type is constructed from a type descriptor that doesn't provide a
+        # name. If it exists, the structure has been anonymously declared; so we check
+        # that the provided argument matches the "shape" of the anonymous structure. If
+        # it matches, modify the invoke signature to match the type of the argument that
+        # was actually provided. The first argument to invoke is the block being
+        # invoked, so we can ignore that type hint.
         for i, argtype in enumerate(self.invoke_argtypes[1:]):
-            if getattr(argtype, "__anonymous__", False):
+            if (
+                isinstance(args[i], Structure)
+                and issubclass(argtype, Structure)
+                and getattr(argtype, "__anonymous__", False)
+            ):
                 anon_fields = [f[1] for f in argtype._fields_]
                 arg_fields = [f[1] for f in args[i]._fields_]
                 if anon_fields != arg_fields:

--- a/src/rubicon/objc/api.py
+++ b/src/rubicon/objc/api.py
@@ -2161,10 +2161,23 @@ class ObjCBlock:
             if not self.has_signature:
                 raise ValueError("Cannot use AUTO types for blocks without signatures")
             restype, *argtypes = ctypes_for_method_encoding(self.signature)
-        self.struct.contents.invoke.restype = ctype_for_type(restype)
-        self.struct.contents.invoke.argtypes = (objc_id,) + tuple(
+            # If the argtypes have been derived from the signature, they will include
+            # the block as the first argument.
+            block_arg = []
+        else:
+            # If the argtypes are explicitly provided, they *won't* include the
+            # first required argument - the block itself.
+            block_arg = [objc_id]
+
+        # Setting the restype and argtypes on the invoke function that is in the
+        # ObjCBlockStruct won't stick, because it's not a distinct Python object that
+        # ctypes can use to attach a type hint. Store the ctypes annotations, and apply
+        # them just before invocation. We're going to have to do some light type
+        # conversion in some cases anyway, so this works out well.
+        self.invoke_restype = ctype_for_type(restype)
+        self.invoke_argtypes = block_arg + [
             ctype_for_type(arg_type) for arg_type in argtypes
-        )
+        ]
 
     def __repr__(self):
         representation = f"<ObjCBlock@{hex(addressof(self.pointer))}"
@@ -2182,7 +2195,35 @@ class ObjCBlock:
         objects according to the default ``ctypes`` rules, based on the
         block's return and parameter types.
         """
-        return self.struct.contents.invoke(self.pointer, *args)
+        # If any of the arguments are structures, they may be anonymous - that is, we
+        # have a descriptor like "{=ii}", which tells us there are two integer fields,
+        # but doesn't provide a name for the structure. ctypes looks for an exact match
+        # of type names, so even if the field types of a structure provided as an
+        # argument match, ctypes will raise a TypeError.
+        #
+        # To avoid this, look for the `__anonymous__` property on the structure
+        # definition. If it exists, check that the provided argument matches the "shape"
+        # of the anonymous structure. If it matches, modify the invoke signature to
+        # match the type of the argument that was actually provided. The first argument
+        # to invoke is the block being invoked, so we can ignore that type hint.
+        for i, argtype in enumerate(self.invoke_argtypes[1:]):
+            if hasattr(argtype, "__anonymous__"):
+                anon_fields = [f[1] for f in argtype._fields_]
+                arg_fields = [f[1] for f in args[i]._fields_]
+                if anon_fields != arg_fields:
+                    raise TypeError(
+                        f"Expected structure with field types {anon_fields} "
+                        f"for argument {i+1}; got {type(args[i]).__name__} "
+                        f"with field types {arg_fields}"
+                    )
+                self.invoke_argtypes[i + 1] = type(args[i])
+
+        # Apply the ctypes hints to the invoke function for the block.
+        invoke = self.struct.contents.invoke
+        invoke.restype = self.invoke_restype
+        invoke.argtypes = self.invoke_argtypes
+
+        return invoke(self.pointer, *args)
 
 
 class ObjCBlockInstance(ObjCInstance):
@@ -2274,8 +2315,7 @@ class Block:
         signature = tuple(ctype_for_type(tp) for tp in argtypes)
 
         restype = ctype_for_type(restype)
-
-        self.cfunc_type = CFUNCTYPE(restype, c_void_p, *signature)
+        cfunc_type = CFUNCTYPE(restype, c_void_p, *signature)
 
         self.literal = BlockLiteral()
         self.literal.isa = addressof(_NSConcreteStackBlock)
@@ -2285,8 +2325,9 @@ class Block:
             | BlockConsts.HAS_COPY_DISPOSE
         )
         self.literal.reserved = 0
-        self.cfunc_wrapper = self.cfunc_type(self.wrapper)
-        self.literal.invoke = cast(self.cfunc_wrapper, c_void_p)
+        cfunc_wrapper = cfunc_type(self.wrapper)
+        self.literal.invoke = cast(cfunc_wrapper, c_void_p)
+
         self.descriptor = BlockDescriptor()
         self.descriptor.reserved = 0
         self.descriptor.size = sizeof(BlockLiteral)
@@ -2305,7 +2346,10 @@ class Block:
         self.block = cast(byref(self.literal), objc_block)
         self._as_parameter_ = self.block
 
-    def wrapper(self, instance, *args):
+    def wrapper(self, block, *args):
+        # ObjC blocks take the block as the first argument when they're invoked;
+        # but since this is a wrapper around a Python object, we know the function
+        # that has to be invoked.
         return self.func(*args)
 
     def dispose_helper(self, dst):

--- a/src/rubicon/objc/api.py
+++ b/src/rubicon/objc/api.py
@@ -2207,7 +2207,7 @@ class ObjCBlock:
         # match the type of the argument that was actually provided. The first argument
         # to invoke is the block being invoked, so we can ignore that type hint.
         for i, argtype in enumerate(self.invoke_argtypes[1:]):
-            if hasattr(argtype, "__anonymous__"):
+            if getattr(argtype, "__anonymous__", False):
                 anon_fields = [f[1] for f in argtype._fields_]
                 arg_fields = [f[1] for f in args[i]._fields_]
                 if anon_fields != arg_fields:

--- a/src/rubicon/objc/types.py
+++ b/src/rubicon/objc/types.py
@@ -259,7 +259,7 @@ def _create_structish_type_for_encoding(encoding, *, base):
     # itself are typed correctly. Also annotate the structure with an `__anonymous__`
     # marker so that we can easily identify anonymous structures.
     py_name = "_Anonymous" if name is None else name.decode("utf-8")
-    structish_type = type(py_name, (base,), {"__anonymous__": True})
+    structish_type = type(py_name, (base,), {"__anonymous__": name is None})
     # Register the structish for its own encoding, so the same type is used in the future.
     register_encoding(encoding, structish_type)
     if name is not None:

--- a/src/rubicon/objc/types.py
+++ b/src/rubicon/objc/types.py
@@ -254,10 +254,12 @@ def _create_structish_type_for_encoding(encoding, *, base):
         # Anonymous structish, has no name.
         name = None
 
-    # Create the subclass. The _fields_ are not set yet, this is done later.
-    # The structish is already registered here, so that pointers to this structish type in itself are typed correctly.
+    # Create the subclass. The _fields_ are not set yet, this is done later. The
+    # structish is already registered here, so that pointers to this structish type in
+    # itself are typed correctly. Also annotate the structure with an `__anonymous__`
+    # marker so that we can easily identify anonymous structures.
     py_name = "_Anonymous" if name is None else name.decode("utf-8")
-    structish_type = type(py_name, (base,), {})
+    structish_type = type(py_name, (base,), {"__anonymous__": True})
     # Register the structish for its own encoding, so the same type is used in the future.
     register_encoding(encoding, structish_type)
     if name is not None:

--- a/tests/test_blocks.py
+++ b/tests/test_blocks.py
@@ -1,7 +1,7 @@
 from __future__ import annotations
 
 import unittest
-from ctypes import Structure, c_int, c_void_p
+from ctypes import Structure, c_float, c_int, c_void_p
 
 from rubicon.objc import NSObject, ObjCBlock, ObjCClass, objc_method
 from rubicon.objc.api import Block
@@ -92,6 +92,32 @@ class BlockTests(unittest.TestCase):
         instance = BlockObjectExample.alloc().initWithDelegate_(delegate)
         result = instance.structBlockExample()
         self.assertEqual(result, 85)
+
+    def test_block_delegate_auto_struct_mismatch(self):
+        class BadBlockStruct(Structure):
+            _fields_ = [
+                ("a", c_float),
+                ("b", c_int),
+            ]
+
+        class BadDelegateAutoStruct(NSObject):
+            @objc_method
+            def structBlockMethod_(self, block: objc_block) -> int:
+                try:
+                    # block accepts an anonymous structure with 2 int arguments
+                    # Passing in BadBlockStruct should raise a type error because
+                    # the structure's shape doesn't match.
+                    return block(BadBlockStruct(2.71828, 43))
+                except TypeError:
+                    # Ideally, this would be raised as an ObjC error;
+                    # however, at least for now, that doesn't happen.
+                    return -99
+
+        BlockObjectExample = ObjCClass("BlockObjectExample")
+        delegate = BadDelegateAutoStruct.alloc().init()
+        instance = BlockObjectExample.alloc().initWithDelegate_(delegate)
+        result = instance.structBlockExample()
+        self.assertEqual(result, -99)
 
     def test_block_receiver(self):
         BlockReceiverExample = ObjCClass("BlockReceiverExample")

--- a/tests/test_blocks.py
+++ b/tests/test_blocks.py
@@ -1,6 +1,5 @@
 from __future__ import annotations
 
-import platform
 import unittest
 from ctypes import Structure, c_int, c_void_p
 
@@ -9,31 +8,19 @@ from rubicon.objc.api import Block
 from rubicon.objc.runtime import objc_block
 
 
-def m1_block_failure(test):
-    # Properties/methods returning blocks currently fail on ARM hardware.
-    # See #225 for details
-    if platform.processor().startswith("arm"):
-        return unittest.expectedFailure(test)
-    else:
-        return test
-
-
 class BlockTests(unittest.TestCase):
-    @m1_block_failure
     def test_block_property_ctypes(self):
         BlockPropertyExample = ObjCClass("BlockPropertyExample")
         instance = BlockPropertyExample.alloc().init()
         result = ObjCBlock(instance.blockProperty, c_int, c_int, c_int)(1, 2)
         self.assertEqual(result, 3)
 
-    @m1_block_failure
     def test_block_property_pytypes(self):
         BlockPropertyExample = ObjCClass("BlockPropertyExample")
         instance = BlockPropertyExample.alloc().init()
         result = ObjCBlock(instance.blockProperty, int, int, int)(1, 2)
         self.assertEqual(result, 3)
 
-    @m1_block_failure
     def test_block_delegate_method_manual_ctypes(self):
         class DelegateManualC(NSObject):
             @objc_method
@@ -46,7 +33,6 @@ class BlockTests(unittest.TestCase):
         result = instance.blockExample()
         self.assertEqual(result, 5)
 
-    @m1_block_failure
     def test_block_delegate_method_manual_pytypes(self):
         class DelegateManualPY(NSObject):
             @objc_method
@@ -59,7 +45,6 @@ class BlockTests(unittest.TestCase):
         result = instance.blockExample()
         self.assertEqual(result, 5)
 
-    @m1_block_failure
     def test_block_delegate_auto(self):
         class DelegateAuto(NSObject):
             @objc_method
@@ -72,7 +57,6 @@ class BlockTests(unittest.TestCase):
         result = instance.blockExample()
         self.assertEqual(result, 9)
 
-    @m1_block_failure
     def test_block_delegate_auto_struct(self):
         class BlockStruct(Structure):
             _fields_ = [
@@ -143,7 +127,6 @@ class BlockTests(unittest.TestCase):
 
         self.assertEqual(values, [27])
 
-    @m1_block_failure
     def test_block_round_trip(self):
         BlockRoundTrip = ObjCClass("BlockRoundTrip")
         instance = BlockRoundTrip.alloc().init()
@@ -172,7 +155,6 @@ class BlockTests(unittest.TestCase):
         returned_block_2 = instance.roundTripNoArgs(block_2)
         self.assertEqual(returned_block_2(), 42)
 
-    @m1_block_failure
     def test_block_bound_method(self):
         """A bound method with type annotations can be wrapped in a block."""
 

--- a/tests/test_blocks.py
+++ b/tests/test_blocks.py
@@ -57,6 +57,24 @@ class BlockTests(unittest.TestCase):
         result = instance.blockExample()
         self.assertEqual(result, 9)
 
+    def test_block_delegate_manual_struct(self):
+        class BlockStruct(Structure):
+            _fields_ = [
+                ("a", c_int),
+                ("b", c_int),
+            ]
+
+        class DelegateManualStruct(NSObject):
+            @objc_method
+            def structBlockMethod_(self, block: objc_block) -> int:
+                return ObjCBlock(block, int, BlockStruct)(BlockStruct(42, 43))
+
+        BlockObjectExample = ObjCClass("BlockObjectExample")
+        delegate = DelegateManualStruct.alloc().init()
+        instance = BlockObjectExample.alloc().initWithDelegate_(delegate)
+        result = instance.structBlockExample()
+        self.assertEqual(result, 85)
+
     def test_block_delegate_auto_struct(self):
         class BlockStruct(Structure):
             _fields_ = [

